// @carbon/react/lib/components/Tile/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var FeatureFlags = require('@carbon/feature-flags');
var Tile$1 = require('./next/Tile.js');
var Tile$2 = require('./Tile.js');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var FeatureFlags__namespace = /*#__PURE__*/_interopNamespace(FeatureFlags);

var Tile = FeatureFlags__namespace.enabled('enable-v11-release') ? Tile$1.Tile : Tile$2.Tile;
var ClickableTile = FeatureFlags__namespace.enabled('enable-v11-release') ? Tile$1.ClickableTile : Tile$2.ClickableTile;
var SelectableTile = FeatureFlags__namespace.enabled('enable-v11-release') ? Tile$1.SelectableTile : Tile$2.SelectableTile;
var ExpandableTile = FeatureFlags__namespace.enabled('enable-v11-release') ? Tile$1.ExpandableTile : Tile$2.ExpandableTile;
var TileAboveTheFoldContent = FeatureFlags__namespace.enabled('enable-v11-release') ? Tile$1.TileAboveTheFoldContent : Tile$2.TileAboveTheFoldContent;
var TileBelowTheFoldContent = FeatureFlags__namespace.enabled('enable-v11-release') ? Tile$1.TileBelowTheFoldContent : Tile$2.TileBelowTheFoldContent;

exports.ClickableTile = ClickableTile;
exports.ExpandableTile = ExpandableTile;
exports.SelectableTile = SelectableTile;
exports.Tile = Tile;
exports.TileAboveTheFoldContent = TileAboveTheFoldContent;
exports.TileBelowTheFoldContent = TileBelowTheFoldContent;
