// @carbon/react/lib/components/Menu/MenuSelectableItem.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var PropTypes = require('prop-types');
var iconsReact = require('@carbon/icons-react');
var MenuOption = require('./MenuOption.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

function MenuSelectableItem(_ref) {
  var label = _ref.label,
      initialChecked = _ref.initialChecked,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? function () {} : _ref$onChange;

  var _useState = React.useState(initialChecked),
      _useState2 = _rollupPluginBabelHelpers.slicedToArray(_useState, 2),
      checked = _useState2[0],
      setChecked = _useState2[1];

  function handleClick() {
    setChecked(!checked);
    onChange(!checked);
  }

  return /*#__PURE__*/React__default["default"].createElement(MenuOption["default"], {
    role: "menuitemcheckbox",
    "aria-checked": checked,
    renderIcon: checked ? iconsReact.Checkmark : null,
    label: label,
    indented: true,
    onClick: handleClick
  });
}

MenuSelectableItem.propTypes = {
  /**
   * Whether the option should be checked by default
   */
  initialChecked: PropTypes__default["default"].bool,

  /**
   * Rendered label for the MenuOptionContent
   */
  label: PropTypes__default["default"].node.isRequired,

  /**
   * Callback function when selection the has been changed
   */
  onChange: PropTypes__default["default"].func
};

exports["default"] = MenuSelectableItem;
