// @carbon/react/es/components/Notification/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { NotificationActionButton as NotificationActionButton$1, NotificationButton as NotificationButton$1, ToastNotification as ToastNotification$1, InlineNotification as InlineNotification$1, ActionableNotification as ActionableNotification$1 } from './next/Notification.js';
import { NotificationActionButton as NotificationActionButton$2, NotificationTextDetails as NotificationTextDetails$1, NotificationButton as NotificationButton$2, ToastNotification as ToastNotification$2, InlineNotification as InlineNotification$2 } from './Notification.js';
import { createComponentToggle } from '../../internal/ComponentToggle.js';

var NotificationActionButton = createComponentToggle({
  name: 'NotificationActionButton',
  next: NotificationActionButton$1,
  classic: NotificationActionButton$2
});
var NotificationTextDetails = createComponentToggle({
  name: 'NotificationTextDetails',
  classic: NotificationTextDetails$1
});
var NotificationButton = createComponentToggle({
  name: 'NotificationButton',
  next: NotificationButton$1,
  classic: NotificationButton$2
});
var ToastNotification = createComponentToggle({
  name: 'ToastNotification',
  next: ToastNotification$1,
  classic: ToastNotification$2
});
var InlineNotification = createComponentToggle({
  name: 'InlineNotification',
  next: InlineNotification$1,
  classic: InlineNotification$2
});
var ActionableNotification = createComponentToggle({
  name: 'ActionableNotification',
  next: ActionableNotification$1,
  classic: null
});

export { ActionableNotification, InlineNotification, NotificationActionButton, NotificationButton, NotificationTextDetails, ToastNotification };
