// @carbon/react/lib/components/Layer/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var PropTypes = require('prop-types');
var cx = require('classnames');
var usePrefix = require('../../internal/usePrefix.js');
var LayerContext = require('./LayerContext.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

var _excluded = ["as", "className", "children", "level"];
var levels = ['one', 'two', 'three'];
var MAX_LEVEL = levels.length - 1;
function Layer(_ref) {
  var _ref$as = _ref.as,
      BaseComponent = _ref$as === void 0 ? 'div' : _ref$as,
      customClassName = _ref.className,
      children = _ref.children,
      overrideLevel = _ref.level,
      rest = _rollupPluginBabelHelpers.objectWithoutProperties(_ref, _excluded);

  var contextLevel = React__default["default"].useContext(LayerContext.LayerContext);
  var level = overrideLevel !== null && overrideLevel !== void 0 ? overrideLevel : contextLevel;
  var prefix = usePrefix.usePrefix();
  var className = cx__default["default"]("".concat(prefix, "--layer-").concat(levels[level]), customClassName); // The level should be between 0 and MAX_LEVEL

  var value = Math.max(0, Math.min(level + 1, MAX_LEVEL));
  return /*#__PURE__*/React__default["default"].createElement(LayerContext.LayerContext.Provider, {
    value: value
  }, /*#__PURE__*/React__default["default"].createElement(BaseComponent, _rollupPluginBabelHelpers["extends"]({}, rest, {
    className: className
  }), children));
}
Layer.propTypes = {
  /**
   * Specify a custom component or element to be rendered as the top-level
   * element in the component
   */
  as: PropTypes__default["default"].oneOfType([PropTypes__default["default"].func, PropTypes__default["default"].string, PropTypes__default["default"].elementType]),

  /**
   * Provide child elements to be rendered inside of `Theme`
   */
  children: PropTypes__default["default"].node,

  /**
   * Provide a custom class name to be used on the outermost element rendered by
   * the component
   */
  className: PropTypes__default["default"].string,

  /**
   * Specify the layer level and override any existing levels based on hierarchy
   */
  level: PropTypes__default["default"].oneOf([0, 1, 2])
};

exports.Layer = Layer;
