// @carbon/react/lib/components/ComposedModal/next/ModalHeader.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var PropTypes = require('prop-types');
var cx = require('classnames');
var iconsReact = require('@carbon/icons-react');
var usePrefix = require('../../../internal/usePrefix.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

var _excluded = ["buttonOnClick", "children", "className", "closeClassName", "closeIconClassName", "closeModal", "iconDescription", "label", "labelClassName", "title", "titleClassName"];
var ModalHeader = /*#__PURE__*/React__default["default"].forwardRef(function ModalHeader(_ref, ref) {
  var _cx, _cx2, _cx3, _cx4, _cx5;

  var buttonOnClick = _ref.buttonOnClick,
      children = _ref.children,
      customClassName = _ref.className,
      closeClassName = _ref.closeClassName,
      closeIconClassName = _ref.closeIconClassName,
      closeModal = _ref.closeModal,
      iconDescription = _ref.iconDescription,
      label = _ref.label,
      labelClassName = _ref.labelClassName,
      title = _ref.title,
      titleClassName = _ref.titleClassName,
      rest = _rollupPluginBabelHelpers.objectWithoutProperties(_ref, _excluded);

  var prefix = usePrefix.usePrefix();

  function handleCloseButtonClick(evt) {
    closeModal(evt);
    buttonOnClick();
  }

  var headerClass = cx__default["default"]((_cx = {}, _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--modal-header"), true), _rollupPluginBabelHelpers.defineProperty(_cx, customClassName, customClassName), _cx));
  var labelClass = cx__default["default"]((_cx2 = {}, _rollupPluginBabelHelpers.defineProperty(_cx2, "".concat(prefix, "--modal-header__label ").concat(prefix, "--type-delta"), true), _rollupPluginBabelHelpers.defineProperty(_cx2, labelClassName, labelClassName), _cx2));
  var titleClass = cx__default["default"]((_cx3 = {}, _rollupPluginBabelHelpers.defineProperty(_cx3, "".concat(prefix, "--modal-header__heading ").concat(prefix, "--type-beta"), true), _rollupPluginBabelHelpers.defineProperty(_cx3, titleClassName, titleClassName), _cx3));
  var closeClass = cx__default["default"]((_cx4 = {}, _rollupPluginBabelHelpers.defineProperty(_cx4, "".concat(prefix, "--modal-close"), true), _rollupPluginBabelHelpers.defineProperty(_cx4, closeClassName, closeClassName), _cx4));
  var closeIconClass = cx__default["default"]((_cx5 = {}, _rollupPluginBabelHelpers.defineProperty(_cx5, "".concat(prefix, "--modal-close__icon"), true), _rollupPluginBabelHelpers.defineProperty(_cx5, closeIconClassName, closeIconClassName), _cx5));
  return /*#__PURE__*/React__default["default"].createElement("div", _rollupPluginBabelHelpers["extends"]({
    className: headerClass
  }, rest, {
    ref: ref
  }), label && /*#__PURE__*/React__default["default"].createElement("h2", {
    className: labelClass
  }, label), title && /*#__PURE__*/React__default["default"].createElement("h3", {
    className: titleClass
  }, title), children, /*#__PURE__*/React__default["default"].createElement("button", {
    onClick: handleCloseButtonClick,
    className: closeClass,
    title: iconDescription,
    "aria-label": iconDescription,
    type: "button"
  }, /*#__PURE__*/React__default["default"].createElement(iconsReact.Close, {
    size: 20,
    className: closeIconClass
  })));
});
ModalHeader.propTypes = {
  /**
   * Provide an optional function to be called when the close button is
   * clicked
   */
  buttonOnClick: PropTypes__default["default"].func,

  /**
   * Specify the content to be placed in the ModalHeader
   */
  children: PropTypes__default["default"].node,

  /**
   * Specify an optional className to be applied to the modal header
   */
  className: PropTypes__default["default"].string,

  /**
   * Specify an optional className to be applied to the modal close node
   */
  closeClassName: PropTypes__default["default"].string,

  /**
   * Specify an optional className to be applied to the modal close icon node
   */
  closeIconClassName: PropTypes__default["default"].string,

  /**
   * Provide an optional function to be called when the modal is closed
   */
  closeModal: PropTypes__default["default"].func,

  /**
   * Specify a description for the close icon that can be read by screen
   * readers
   */
  iconDescription: PropTypes__default["default"].string,

  /**
   * Specify an optional label to be displayed
   */
  label: PropTypes__default["default"].node,

  /**
   * Specify an optional className to be applied to the modal header label
   */
  labelClassName: PropTypes__default["default"].string,

  /**
   * Specify an optional title to be displayed
   */
  title: PropTypes__default["default"].node,

  /**
   * Specify an optional className to be applied to the modal heading
   */
  titleClassName: PropTypes__default["default"].string
};
ModalHeader.defaultProps = {
  iconDescription: 'Close',
  buttonOnClick: function buttonOnClick() {},
  closeModal: function closeModal() {}
};

exports.ModalHeader = ModalHeader;
