// @carbon/react/es/components/Tile/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as FeatureFlags from '@carbon/feature-flags';
import { Tile as Tile$1, ClickableTile as ClickableTile$1, SelectableTile as SelectableTile$1, ExpandableTile as ExpandableTile$1, TileAboveTheFoldContent as TileAboveTheFoldContent$1, TileBelowTheFoldContent as TileBelowTheFoldContent$1 } from './next/Tile.js';
import { Tile as Tile$2, ClickableTile as ClickableTile$2, SelectableTile as SelectableTile$2, ExpandableTile as ExpandableTile$2, TileAboveTheFoldContent as TileAboveTheFoldContent$2, TileBelowTheFoldContent as TileBelowTheFoldContent$2 } from './Tile.js';

var Tile = FeatureFlags.enabled('enable-v11-release') ? Tile$1 : Tile$2;
var ClickableTile = FeatureFlags.enabled('enable-v11-release') ? ClickableTile$1 : ClickableTile$2;
var SelectableTile = FeatureFlags.enabled('enable-v11-release') ? SelectableTile$1 : SelectableTile$2;
var ExpandableTile = FeatureFlags.enabled('enable-v11-release') ? ExpandableTile$1 : ExpandableTile$2;
var TileAboveTheFoldContent = FeatureFlags.enabled('enable-v11-release') ? TileAboveTheFoldContent$1 : TileAboveTheFoldContent$2;
var TileBelowTheFoldContent = FeatureFlags.enabled('enable-v11-release') ? TileBelowTheFoldContent$1 : TileBelowTheFoldContent$2;

export { ClickableTile, ExpandableTile, SelectableTile, Tile, TileAboveTheFoldContent, TileBelowTheFoldContent };
