// @carbon/react/lib/components/RadioButtonGroup/next/RadioButtonGroup.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../../_virtual/_rollupPluginBabelHelpers.js');
var PropTypes = require('prop-types');
var React = require('react');
var cx = require('classnames');
var index = require('../../Text/index.js');
var usePrefix = require('../../../internal/usePrefix.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

var RadioButtonGroup = /*#__PURE__*/React__default["default"].forwardRef(function RadioButtonGroup(_ref, ref) {
  var _classNames;

  var children = _ref.children,
      className = _ref.className,
      defaultSelected = _ref.defaultSelected,
      disabled = _ref.disabled,
      _ref$labelPosition = _ref.labelPosition,
      labelPosition = _ref$labelPosition === void 0 ? 'right' : _ref$labelPosition,
      legendText = _ref.legendText,
      name = _ref.name,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? function () {} : _ref$onChange,
      _ref$orientation = _ref.orientation,
      orientation = _ref$orientation === void 0 ? 'horizontal' : _ref$orientation,
      valueSelected = _ref.valueSelected;
  var prefix = usePrefix.usePrefix();

  var _useState = React.useState(valueSelected !== null && valueSelected !== void 0 ? valueSelected : defaultSelected),
      _useState2 = _rollupPluginBabelHelpers.slicedToArray(_useState, 2),
      selected = _useState2[0],
      setSelected = _useState2[1];

  var _useState3 = React.useState(valueSelected),
      _useState4 = _rollupPluginBabelHelpers.slicedToArray(_useState3, 2),
      prevValueSelected = _useState4[0],
      setPrevValueSelected = _useState4[1];
  /**
   * prop + state alignment - getDerivedStateFromProps
   * only update if selected prop changes
   */


  if (valueSelected !== prevValueSelected) {
    setSelected(valueSelected);
    setPrevValueSelected(valueSelected);
  }

  function getRadioButtons() {
    var mappedChildren = React__default["default"].Children.map(children, function (radioButton) {
      var value = radioButton.props.value;
      return /*#__PURE__*/React__default["default"].cloneElement(radioButton, {
        name: name,
        key: value,
        value: value,
        onChange: handleOnChange,
        checked: value === selected
      });
    });
    return mappedChildren;
  }

  function handleOnChange(newSelection, value, evt) {
    if (newSelection !== selected) {
      setSelected(newSelection);
      onChange(newSelection, name, evt);
    }
  }

  var fieldsetClasses = cx__default["default"]("".concat(prefix, "--radio-button-group"), (_classNames = {}, _rollupPluginBabelHelpers.defineProperty(_classNames, "".concat(prefix, "--radio-button-group--").concat(orientation), orientation === 'vertical'), _rollupPluginBabelHelpers.defineProperty(_classNames, "".concat(prefix, "--radio-button-group--label-").concat(labelPosition), labelPosition), _classNames));
  var wrapperClasses = cx__default["default"]("".concat(prefix, "--form-item"), className);
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: wrapperClasses,
    ref: ref
  }, /*#__PURE__*/React__default["default"].createElement("fieldset", {
    className: fieldsetClasses,
    disabled: disabled
  }, legendText && /*#__PURE__*/React__default["default"].createElement(index.Legend, {
    className: "".concat(prefix, "--label")
  }, legendText), getRadioButtons()));
});
RadioButtonGroup.propTypes = {
  /**
   * Provide a collection of `<RadioButton>` components to render in the group
   */
  children: PropTypes__default["default"].node,

  /**
   * Provide an optional className to be applied to the container node
   */
  className: PropTypes__default["default"].string,

  /**
   * Specify the `<RadioButton>` to be selected by default
   */
  defaultSelected: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),

  /**
   * Specify whether the group is disabled
   */
  disabled: PropTypes__default["default"].bool,

  /**
   * Provide where label text should be placed
   */
  labelPosition: PropTypes__default["default"].oneOf(['left', 'right']),

  /**
   * Provide a legend to the RadioButtonGroup input that you are
   * exposing to the user
   */
  legendText: PropTypes__default["default"].node,

  /**
   * Specify the name of the underlying `<input>` nodes
   */
  name: PropTypes__default["default"].string.isRequired,

  /**
   * Provide an optional `onChange` hook that is called whenever the value of
   * the group changes
   */
  onChange: PropTypes__default["default"].func,

  /**
   * Provide where radio buttons should be placed
   */
  orientation: PropTypes__default["default"].oneOf(['horizontal', 'vertical']),

  /**
   * Specify the value that is currently selected in the group
   */
  valueSelected: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number])
};
var RadioButtonGroupNext = RadioButtonGroup;

exports["default"] = RadioButtonGroupNext;
