// @slippi/slippi-js/dist/melee/types.esm.js
var Character;

(function (Character) {
  Character[Character["CAPTAIN_FALCON"] = 0] = "CAPTAIN_FALCON";
  Character[Character["DONKEY_KONG"] = 1] = "DONKEY_KONG";
  Character[Character["FOX"] = 2] = "FOX";
  Character[Character["GAME_AND_WATCH"] = 3] = "GAME_AND_WATCH";
  Character[Character["KIRBY"] = 4] = "KIRBY";
  Character[Character["BOWSER"] = 5] = "BOWSER";
  Character[Character["LINK"] = 6] = "LINK";
  Character[Character["LUIGI"] = 7] = "LUIGI";
  Character[Character["MARIO"] = 8] = "MARIO";
  Character[Character["MARTH"] = 9] = "MARTH";
  Character[Character["MEWTWO"] = 10] = "MEWTWO";
  Character[Character["NESS"] = 11] = "NESS";
  Character[Character["PEACH"] = 12] = "PEACH";
  Character[Character["PIKACHU"] = 13] = "PIKACHU";
  Character[Character["ICE_CLIMBERS"] = 14] = "ICE_CLIMBERS";
  Character[Character["JIGGLYPUFF"] = 15] = "JIGGLYPUFF";
  Character[Character["SAMUS"] = 16] = "SAMUS";
  Character[Character["YOSHI"] = 17] = "YOSHI";
  Character[Character["ZELDA"] = 18] = "ZELDA";
  Character[Character["SHEIK"] = 19] = "SHEIK";
  Character[Character["FALCO"] = 20] = "FALCO";
  Character[Character["YOUNG_LINK"] = 21] = "YOUNG_LINK";
  Character[Character["DR_MARIO"] = 22] = "DR_MARIO";
  Character[Character["ROY"] = 23] = "ROY";
  Character[Character["PICHU"] = 24] = "PICHU";
  Character[Character["GANONDORF"] = 25] = "GANONDORF";
  Character[Character["MASTER_HAND"] = 26] = "MASTER_HAND";
  Character[Character["WIREFRAME_MALE"] = 27] = "WIREFRAME_MALE";
  Character[Character["WIREFRAME_FEMALE"] = 28] = "WIREFRAME_FEMALE";
  Character[Character["GIGA_BOWSER"] = 29] = "GIGA_BOWSER";
  Character[Character["CRAZY_HAND"] = 30] = "CRAZY_HAND";
  Character[Character["SANDBAG"] = 31] = "SANDBAG";
  Character[Character["POPO"] = 32] = "POPO";
})(Character || (Character = {}));

var Stage;

(function (Stage) {
  Stage[Stage["FOUNTAIN_OF_DREAMS"] = 2] = "FOUNTAIN_OF_DREAMS";
  Stage[Stage["POKEMON_STADIUM"] = 3] = "POKEMON_STADIUM";
  Stage[Stage["PEACHS_CASTLE"] = 4] = "PEACHS_CASTLE";
  Stage[Stage["KONGO_JUNGLE"] = 5] = "KONGO_JUNGLE";
  Stage[Stage["BRINSTAR"] = 6] = "BRINSTAR";
  Stage[Stage["CORNERIA"] = 7] = "CORNERIA";
  Stage[Stage["YOSHIS_STORY"] = 8] = "YOSHIS_STORY";
  Stage[Stage["ONETT"] = 9] = "ONETT";
  Stage[Stage["MUTE_CITY"] = 10] = "MUTE_CITY";
  Stage[Stage["RAINBOW_CRUISE"] = 11] = "RAINBOW_CRUISE";
  Stage[Stage["JUNGLE_JAPES"] = 12] = "JUNGLE_JAPES";
  Stage[Stage["GREAT_BAY"] = 13] = "GREAT_BAY";
  Stage[Stage["HYRULE_TEMPLE"] = 14] = "HYRULE_TEMPLE";
  Stage[Stage["BRINSTAR_DEPTHS"] = 15] = "BRINSTAR_DEPTHS";
  Stage[Stage["YOSHIS_ISLAND"] = 16] = "YOSHIS_ISLAND";
  Stage[Stage["GREEN_GREENS"] = 17] = "GREEN_GREENS";
  Stage[Stage["FOURSIDE"] = 18] = "FOURSIDE";
  Stage[Stage["MUSHROOM_KINGDOM"] = 19] = "MUSHROOM_KINGDOM";
  Stage[Stage["MUSHROOM_KINGDOM_2"] = 20] = "MUSHROOM_KINGDOM_2";
  Stage[Stage["VENOM"] = 22] = "VENOM";
  Stage[Stage["POKE_FLOATS"] = 23] = "POKE_FLOATS";
  Stage[Stage["BIG_BLUE"] = 24] = "BIG_BLUE";
  Stage[Stage["ICICLE_MOUNTAIN"] = 25] = "ICICLE_MOUNTAIN";
  Stage[Stage["ICETOP"] = 26] = "ICETOP";
  Stage[Stage["FLAT_ZONE"] = 27] = "FLAT_ZONE";
  Stage[Stage["DREAMLAND"] = 28] = "DREAMLAND";
  Stage[Stage["YOSHIS_ISLAND_N64"] = 29] = "YOSHIS_ISLAND_N64";
  Stage[Stage["KONGO_JUNGLE_N64"] = 30] = "KONGO_JUNGLE_N64";
  Stage[Stage["BATTLEFIELD"] = 31] = "BATTLEFIELD";
  Stage[Stage["FINAL_DESTINATION"] = 32] = "FINAL_DESTINATION";
  Stage[Stage["TARGET_TEST_MARIO"] = 33] = "TARGET_TEST_MARIO";
  Stage[Stage["TARGET_TEST_CAPTAIN_FALCON"] = 34] = "TARGET_TEST_CAPTAIN_FALCON";
  Stage[Stage["TARGET_TEST_YOUNG_LINK"] = 35] = "TARGET_TEST_YOUNG_LINK";
  Stage[Stage["TARGET_TEST_DONKEY_KONG"] = 36] = "TARGET_TEST_DONKEY_KONG";
  Stage[Stage["TARGET_TEST_DR_MARIO"] = 37] = "TARGET_TEST_DR_MARIO";
  Stage[Stage["TARGET_TEST_FALCO"] = 38] = "TARGET_TEST_FALCO";
  Stage[Stage["TARGET_TEST_FOX"] = 39] = "TARGET_TEST_FOX";
  Stage[Stage["TARGET_TEST_ICE_CLIMBERS"] = 40] = "TARGET_TEST_ICE_CLIMBERS";
  Stage[Stage["TARGET_TEST_KIRBY"] = 41] = "TARGET_TEST_KIRBY";
  Stage[Stage["TARGET_TEST_BOWSER"] = 42] = "TARGET_TEST_BOWSER";
  Stage[Stage["TARGET_TEST_LINK"] = 43] = "TARGET_TEST_LINK";
  Stage[Stage["TARGET_TEST_LUIGI"] = 44] = "TARGET_TEST_LUIGI";
  Stage[Stage["TARGET_TEST_MARTH"] = 45] = "TARGET_TEST_MARTH";
  Stage[Stage["TARGET_TEST_MEWTWO"] = 46] = "TARGET_TEST_MEWTWO";
  Stage[Stage["TARGET_TEST_NESS"] = 47] = "TARGET_TEST_NESS";
  Stage[Stage["TARGET_TEST_PEACH"] = 48] = "TARGET_TEST_PEACH";
  Stage[Stage["TARGET_TEST_PICHU"] = 49] = "TARGET_TEST_PICHU";
  Stage[Stage["TARGET_TEST_PIKACHU"] = 50] = "TARGET_TEST_PIKACHU";
  Stage[Stage["TARGET_TEST_JIGGLYPUFF"] = 51] = "TARGET_TEST_JIGGLYPUFF";
  Stage[Stage["TARGET_TEST_SAMUS"] = 52] = "TARGET_TEST_SAMUS";
  Stage[Stage["TARGET_TEST_SHEIK"] = 53] = "TARGET_TEST_SHEIK";
  Stage[Stage["TARGET_TEST_YOSHI"] = 54] = "TARGET_TEST_YOSHI";
  Stage[Stage["TARGET_TEST_ZELDA"] = 55] = "TARGET_TEST_ZELDA";
  Stage[Stage["TARGET_TEST_GAME_AND_WATCH"] = 56] = "TARGET_TEST_GAME_AND_WATCH";
  Stage[Stage["TARGET_TEST_ROY"] = 57] = "TARGET_TEST_ROY";
  Stage[Stage["TARGET_TEST_GANONDORF"] = 58] = "TARGET_TEST_GANONDORF";
  Stage[Stage["HOME_RUN_CONTEST"] = 84] = "HOME_RUN_CONTEST";
})(Stage || (Stage = {}));

export { Character, Stage };
