// @carbon/react/es/components/ToggleSmall/ToggleSmall.Skeleton.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { inherits as _inherits, createSuper as _createSuper, classCallCheck as _classCallCheck, createClass as _createClass, defineProperty as _defineProperty, objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import PropTypes from 'prop-types';
import React__default from 'react';
import cx from 'classnames';
import { PrefixContext } from '../../internal/usePrefix.js';

var _path, _defineProperty2;

var _excluded = ["id", "labelText", "className"];

var ToggleSmallSkeleton = /*#__PURE__*/function (_React$Component) {
  _inherits(ToggleSmallSkeleton, _React$Component);

  var _super = _createSuper(ToggleSmallSkeleton);

  function ToggleSmallSkeleton() {
    _classCallCheck(this, ToggleSmallSkeleton);

    return _super.apply(this, arguments);
  }

  _createClass(ToggleSmallSkeleton, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          id = _this$props.id,
          labelText = _this$props.labelText,
          className = _this$props.className,
          rest = _objectWithoutProperties(_this$props, _excluded);

      return /*#__PURE__*/React__default.createElement(PrefixContext.Consumer, null, function (prefix) {
        return /*#__PURE__*/React__default.createElement("div", _extends({
          className: cx("".concat(prefix, "--form-item"), className)
        }, rest), /*#__PURE__*/React__default.createElement("input", {
          type: "checkbox",
          id: id,
          className: "".concat(prefix, "--toggle ").concat(prefix, "--toggle--small ").concat(prefix, "--skeleton")
        }), /*#__PURE__*/React__default.createElement("label", {
          className: "".concat(prefix, "--toggle__label ").concat(prefix, "--skeleton"),
          htmlFor: id
        }, labelText && /*#__PURE__*/React__default.createElement("span", {
          className: "".concat(prefix, "--toggle__label-text")
        }, labelText), /*#__PURE__*/React__default.createElement("span", {
          className: "".concat(prefix, "--toggle__appearance")
        }, /*#__PURE__*/React__default.createElement("svg", {
          className: "".concat(prefix, "--toggle__check"),
          width: "6px",
          height: "5px",
          viewBox: "0 0 6 5"
        }, _path || (_path = /*#__PURE__*/React__default.createElement("path", {
          d: "M2.2403 2.7299L4.9245 0 6 1.1117 2.2384 5 0 2.6863 1.0612 1.511z"
        }))))));
      });
    }
  }]);

  return ToggleSmallSkeleton;
}(React__default.Component);

_defineProperty(ToggleSmallSkeleton, "propTypes", (_defineProperty2 = {}, _defineProperty(_defineProperty2, 'aria-label', PropTypes.string.isRequired), _defineProperty(_defineProperty2, "className", PropTypes.string), _defineProperty(_defineProperty2, "id", PropTypes.string), _defineProperty(_defineProperty2, "labelText", PropTypes.string), _defineProperty2));

_defineProperty(ToggleSmallSkeleton, "defaultProps", _defineProperty({}, 'aria-label', 'Toggle is loading'));

export { ToggleSmallSkeleton as default };
