// @carbon/react/es/components/StructuredList/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { StructuredListWrapper as StructuredListWrapper$1, StructuredListHead as StructuredListHead$1, StructuredListInput as StructuredListInput$1, StructuredListBody as StructuredListBody$1, StructuredListRow as StructuredListRow$1, StructuredListCell as StructuredListCell$1 } from './next/StructuredList.js';
import { StructuredListWrapper as StructuredListWrapper$2, StructuredListHead as StructuredListHead$2, StructuredListInput as StructuredListInput$2, StructuredListBody as StructuredListBody$2, StructuredListRow as StructuredListRow$2, StructuredListCell as StructuredListCell$2 } from './StructuredList.js';
import { createComponentToggle } from '../../internal/ComponentToggle.js';

var StructuredListWrapper = createComponentToggle({
  name: 'StructuredListWrapper',
  next: StructuredListWrapper$1,
  classic: StructuredListWrapper$2
});
var StructuredListHead = createComponentToggle({
  name: 'StructuredListHead',
  next: StructuredListHead$1,
  classic: StructuredListHead$2
});
var StructuredListInput = createComponentToggle({
  name: 'StructuredListInput',
  next: StructuredListInput$1,
  classic: StructuredListInput$2
});
var StructuredListBody = createComponentToggle({
  name: 'StructuredListBody',
  next: StructuredListBody$1,
  classic: StructuredListBody$2
});
var StructuredListRow = createComponentToggle({
  name: 'StructuredListRow',
  next: StructuredListRow$1,
  classic: StructuredListRow$2
});
var StructuredListCell = createComponentToggle({
  name: 'StructuredListCell',
  next: StructuredListCell$1,
  classic: StructuredListCell$2
});

export { StructuredListBody, StructuredListCell, StructuredListHead, StructuredListInput, StructuredListRow, StructuredListWrapper };
