// @carbon/react@1.42.1/lib/components/Menu/MenuContext.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const menuDefaultState = {
  isRoot: true,
  hasIcons: false,
  size: null,
  items: [],
  requestCloseRoot: () => {}
};
function menuReducer(state, action) {
  switch (action.type) {
    case 'enableIcons':
      return {
        ...state,
        hasIcons: true
      };
    case 'registerItem':
      return {
        ...state,
        items: [...state.items, action.payload].filter(item => item.ref.current !== null)
      };
  }
}
const MenuContext = /*#__PURE__*/React__default["default"].createContext({
  state: menuDefaultState,
  // 'dispatch' is populated by the root menu
  dispatch: () => {}
});

exports.MenuContext = MenuContext;
exports.menuReducer = menuReducer;
