// @carbon/react/lib/components/Search/next/Search.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../../_virtual/_rollupPluginBabelHelpers.js');
var iconsReact = require('@carbon/icons-react');
var cx = require('classnames');
var PropTypes = require('prop-types');
var React = require('react');
var index = require('../../../internal/focus/index.js');
var useId = require('../../../internal/useId.js');
var usePrefix = require('../../../internal/usePrefix.js');
var events = require('../../../tools/events.js');
var useMergedRefs = require('../../../internal/useMergedRefs.js');
var deprecate = require('../../../prop-types/deprecate.js');
var match = require('../../../internal/keyboard/match.js');
var keys = require('../../../internal/keyboard/keys.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var _Close;

var _excluded = ["autoComplete", "className", "closeButtonLabelText", "defaultValue", "disabled", "id", "labelText", "light", "onChange", "onClear", "onKeyDown", "onExpand", "placeholder", "renderIcon", "role", "size", "type", "value"];
var Search = /*#__PURE__*/React__default["default"].forwardRef(function Search(_ref, forwardRef) {
  var _cx, _cx2;

  var _ref$autoComplete = _ref.autoComplete,
      autoComplete = _ref$autoComplete === void 0 ? 'off' : _ref$autoComplete,
      className = _ref.className,
      _ref$closeButtonLabel = _ref.closeButtonLabelText,
      closeButtonLabelText = _ref$closeButtonLabel === void 0 ? 'Clear search input' : _ref$closeButtonLabel,
      defaultValue = _ref.defaultValue,
      disabled = _ref.disabled,
      id = _ref.id,
      labelText = _ref.labelText,
      light = _ref.light,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? function () {} : _ref$onChange,
      _ref$onClear = _ref.onClear,
      onClear = _ref$onClear === void 0 ? function () {} : _ref$onClear,
      onKeyDown = _ref.onKeyDown,
      onExpand = _ref.onExpand,
      _ref$placeholder = _ref.placeholder,
      placeholder = _ref$placeholder === void 0 ? '' : _ref$placeholder,
      renderIcon = _ref.renderIcon,
      _ref$role = _ref.role,
      role = _ref$role === void 0 ? 'searchbox' : _ref$role,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'md' : _ref$size,
      _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'text' : _ref$type,
      value = _ref.value,
      rest = _rollupPluginBabelHelpers.objectWithoutProperties(_ref, _excluded);

  var prefix = usePrefix.usePrefix();
  var inputRef = React.useRef(null);
  var ref = useMergedRefs.useMergedRefs([forwardRef, inputRef]);
  var inputId = useId.useId('search-input');
  var uniqueId = id || inputId;
  var searchId = "".concat(uniqueId, "-search");

  var _useState = React.useState(value || defaultValue || false),
      _useState2 = _rollupPluginBabelHelpers.slicedToArray(_useState, 2),
      hasContent = _useState2[0],
      setHasContent = _useState2[1];

  var _useState3 = React.useState(value),
      _useState4 = _rollupPluginBabelHelpers.slicedToArray(_useState3, 2),
      prevValue = _useState4[0],
      setPrevValue = _useState4[1];

  var searchClasses = cx__default["default"]((_cx = {}, _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--search"), true), _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--search--sm"), size === 'sm'), _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--search--md"), size === 'md'), _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--search--lg"), size === 'lg'), _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--search--light"), light), _rollupPluginBabelHelpers.defineProperty(_cx, "".concat(prefix, "--search--disabled"), disabled), _rollupPluginBabelHelpers.defineProperty(_cx, className, className), _cx));
  var clearClasses = cx__default["default"]((_cx2 = {}, _rollupPluginBabelHelpers.defineProperty(_cx2, "".concat(prefix, "--search-close"), true), _rollupPluginBabelHelpers.defineProperty(_cx2, "".concat(prefix, "--search-close--hidden"), !hasContent), _cx2));

  if (value !== prevValue) {
    setHasContent(!!value);
    setPrevValue(value);
  }

  function clearInput(event) {
    if (!value) {
      inputRef.current.value = '';
      onChange(event);
    } else {
      var clearedEvt = Object.assign({}, event.target, {
        target: {
          value: ''
        }
      });
      onChange(clearedEvt);
    }

    onClear();
    setHasContent(false);
    index.focus(inputRef);
  }

  function handleChange(event) {
    setHasContent(event.target.value !== '');
  }

  function handleKeyDown(event) {
    if (match.match(event, keys.Escape)) {
      event.stopPropagation();
      clearInput(event);
    }
  }

  return /*#__PURE__*/React__default["default"].createElement("div", {
    role: "search",
    "aria-labelledby": searchId,
    className: searchClasses
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    role: onExpand ? 'button' : null,
    className: "".concat(prefix, "--search-magnifier"),
    onClick: function onClick() {
      if (onExpand) {
        onExpand();
      }
    }
  }, /*#__PURE__*/React__default["default"].createElement(CustomSearchIcon, {
    icon: renderIcon
  })), /*#__PURE__*/React__default["default"].createElement("label", {
    id: searchId,
    htmlFor: uniqueId,
    className: "".concat(prefix, "--label")
  }, labelText), /*#__PURE__*/React__default["default"].createElement("input", _rollupPluginBabelHelpers["extends"]({}, rest, {
    autoComplete: autoComplete,
    className: "".concat(prefix, "--search-input"),
    defaultValue: defaultValue,
    disabled: disabled,
    role: role,
    ref: ref,
    id: uniqueId,
    onChange: events.composeEventHandlers([onChange, handleChange]),
    onKeyDown: events.composeEventHandlers([onKeyDown, handleKeyDown]),
    placeholder: placeholder,
    type: type,
    value: value
  })), /*#__PURE__*/React__default["default"].createElement("button", {
    "aria-label": closeButtonLabelText,
    className: clearClasses,
    disabled: disabled,
    onClick: clearInput,
    type: "button"
  }, _Close || (_Close = /*#__PURE__*/React__default["default"].createElement(iconsReact.Close, null))));
});
Search.propTypes = {
  /**
   * Specify an optional value for the `autocomplete` property on the underlying
   * `<input>`, defaults to "off"
   */
  autoComplete: PropTypes__default["default"].string,

  /**
   * Specify an optional className to be applied to the container node
   */
  className: PropTypes__default["default"].string,

  /**
   * Specify a label to be read by screen readers on the "close" button
   */
  closeButtonLabelText: PropTypes__default["default"].string,

  /**
   * Optionally provide the default value of the `<input>`
   */
  defaultValue: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),

  /**
   * Specify whether the `<input>` should be disabled
   */
  disabled: PropTypes__default["default"].bool,

  /**
   * Specify a custom `id` for the input
   */
  id: PropTypes__default["default"].string,

  /**
   * Provide the label text for the Search icon
   */
  labelText: PropTypes__default["default"].node.isRequired,

  /**
   * Specify light version or default version of this control
   */
  light: deprecate["default"](PropTypes__default["default"].bool, 'The `light` prop for `Search` is no longer needed and has ' + 'been deprecated in v11 in favor of the new `Layer` component. It will be moved in the next major release.'),

  /**
   * Optional callback called when the search value changes.
   */
  onChange: PropTypes__default["default"].func,

  /**
   * Optional callback called when the search value is cleared.
   */
  onClear: PropTypes__default["default"].func,

  /**
   * Optional callback called when the magnifier icon is clicked in ExpandableSearch.
   */
  onExpand: PropTypes__default["default"].func,

  /**
   * Provide a handler that is invoked on the key down event for the input
   */
  onKeyDown: PropTypes__default["default"].func,

  /**
   * Provide an optional placeholder text for the Search.
   * Note: if the label and placeholder differ,
   * VoiceOver on Mac will read both
   */
  placeholder: PropTypes__default["default"].string,

  /**
   * Rendered icon for the Search.
   * Can be a React component class
   */
  renderIcon: PropTypes__default["default"].oneOfType([PropTypes__default["default"].func, PropTypes__default["default"].object]),

  /**
   * Specify the role for the underlying `<input>`, defaults to `searchbox`
   */
  role: PropTypes__default["default"].string,

  /**
   * Specify the size of the Search
   */
  size: PropTypes__default["default"].oneOf(['sm', 'md', 'lg']),

  /**
   * Optional prop to specify the type of the `<input>`
   */
  type: PropTypes__default["default"].string,

  /**
   * Specify the value of the `<input>`
   */
  value: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number])
};

function CustomSearchIcon(_ref2) {
  var icon = _ref2.icon;
  var prefix = usePrefix.usePrefix();

  if (icon) {
    return /*#__PURE__*/React__default["default"].cloneElement(icon, {
      className: "".concat(prefix, "--search-magnifier-icon")
    });
  }

  return /*#__PURE__*/React__default["default"].createElement(iconsReact.Search, {
    className: "".concat(prefix, "--search-magnifier-icon")
  });
}

CustomSearchIcon.propTypes = {
  /**
   * Rendered icon for the Search. Can be a React component class
   */
  icon: PropTypes__default["default"].oneOfType([PropTypes__default["default"].func, PropTypes__default["default"].object])
};
var SearchNext = Search;

exports["default"] = SearchNext;
