// @carbon/react/lib/components/StructuredList/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var StructuredList = require('./next/StructuredList.js');
var StructuredList$1 = require('./StructuredList.js');
var ComponentToggle = require('../../internal/ComponentToggle.js');

var StructuredListWrapper = ComponentToggle.createComponentToggle({
  name: 'StructuredListWrapper',
  next: StructuredList.StructuredListWrapper,
  classic: StructuredList$1.StructuredListWrapper
});
var StructuredListHead = ComponentToggle.createComponentToggle({
  name: 'StructuredListHead',
  next: StructuredList.StructuredListHead,
  classic: StructuredList$1.StructuredListHead
});
var StructuredListInput = ComponentToggle.createComponentToggle({
  name: 'StructuredListInput',
  next: StructuredList.StructuredListInput,
  classic: StructuredList$1.StructuredListInput
});
var StructuredListBody = ComponentToggle.createComponentToggle({
  name: 'StructuredListBody',
  next: StructuredList.StructuredListBody,
  classic: StructuredList$1.StructuredListBody
});
var StructuredListRow = ComponentToggle.createComponentToggle({
  name: 'StructuredListRow',
  next: StructuredList.StructuredListRow,
  classic: StructuredList$1.StructuredListRow
});
var StructuredListCell = ComponentToggle.createComponentToggle({
  name: 'StructuredListCell',
  next: StructuredList.StructuredListCell,
  classic: StructuredList$1.StructuredListCell
});

exports.StructuredListBody = StructuredListBody;
exports.StructuredListCell = StructuredListCell;
exports.StructuredListHead = StructuredListHead;
exports.StructuredListInput = StructuredListInput;
exports.StructuredListRow = StructuredListRow;
exports.StructuredListWrapper = StructuredListWrapper;
