// @carbon/react/lib/components/Notification/index.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Notification = require('./next/Notification.js');
var Notification$1 = require('./Notification.js');
var ComponentToggle = require('../../internal/ComponentToggle.js');

var NotificationActionButton = ComponentToggle.createComponentToggle({
  name: 'NotificationActionButton',
  next: Notification.NotificationActionButton,
  classic: Notification$1.NotificationActionButton
});
var NotificationTextDetails = ComponentToggle.createComponentToggle({
  name: 'NotificationTextDetails',
  classic: Notification$1.NotificationTextDetails
});
var NotificationButton = ComponentToggle.createComponentToggle({
  name: 'NotificationButton',
  next: Notification.NotificationButton,
  classic: Notification$1.NotificationButton
});
var ToastNotification = ComponentToggle.createComponentToggle({
  name: 'ToastNotification',
  next: Notification.ToastNotification,
  classic: Notification$1.ToastNotification
});
var InlineNotification = ComponentToggle.createComponentToggle({
  name: 'InlineNotification',
  next: Notification.InlineNotification,
  classic: Notification$1.InlineNotification
});
var ActionableNotification = ComponentToggle.createComponentToggle({
  name: 'ActionableNotification',
  next: Notification.ActionableNotification,
  classic: null
});

exports.ActionableNotification = ActionableNotification;
exports.InlineNotification = InlineNotification;
exports.NotificationActionButton = NotificationActionButton;
exports.NotificationButton = NotificationButton;
exports.NotificationTextDetails = NotificationTextDetails;
exports.ToastNotification = ToastNotification;
